### ISSUE TITLE
Handicap training games with uneven playouts

### ISSUE THREAD

**Issue by OmnipotentEntity** [gcp/leela-zero#1531]
I've been working on finding the "50-50" point or at least close enough for reasonably competitive results using a total of 6400 visits divided between two engines for the purpose of setting up a handicapping system for self-play so that Leela can learn how to come from behind and the best way to stay ahead.

Thus far I've found a 50-50 point for 2 handicap games (with 7.5 komi) at around 1300 visits vs 5100 visits.  But the 50-50 point for 3 handicap games is somewhere in the neighborhood of 125 vs 6275.  Furthermore, the 50-50 point drastically changes when turning on noise and randomness.  (It shifts to more even amounts of visits between the engines. It's rather dramatic in the 3 handicap games, less in the 2, though I haven't nailed down exactly where it goes yet.)

Is this something that the project is interested in pursuing?

**Comment by OmnipotentEntity** [comment]
I really think that if there is interest in pursuing this, then a few questions should be answered about exactly how the uneven playouts should work.  Going under 3200 visits is always iffy because it opens the net up to making bad decisions that it would never ever make normally (like ladders), but on the flip side that would mean the other side would need way more visits to keep pace.

Moreover, it's not like 5100 visits is *that* much stronger than 3200, so this might wind up being that the uneven games only function as sorta bad regular training games.

Unfortunately, the particulars of training and overtraining avoidance and so on are not intuitive to me at this time, so I'm hoping that someone better than I am can give their take.

Current project is finding N where 2 handicap games where 3200 visits (black) vs N visits (white) gives near a 50-50 win rate.

**Comment by pcengine** [comment]
I read your comments carefully, and I have a question: Why use uneven playouts to train handicap networks when there are many weaker networks available and those weaker networks might simulate weak side better? 

In a sense, training handicap games with same network using uneven playouts is more or less  like the same player playing slow vs himself playing fast. In real life, a pro player playing slow and carefully still can't give 3 stones against himself playing lightning fast. But any pro player can give H6 or more to kyu players and still win easily. So wouldn't it be better to train it using **Different** (one strong, one weak simulating weak player) networks if your goal is to train the network to play handicap game well? This is just my two cents.

**Comment by OmnipotentEntity** [comment]
The theory that I'm working on, and it may be completely wrong and dumb, is that the difference in visits *might* be more consistent from one network to the next, whereas match ups with different networks would mean hunting down a correctly performing network each time, and verifying that the network assigned to be the weaker net isn't either too strong or too weak.

**Comment by betterworld** [comment]
> Why use uneven playouts to train handicap networks when there are many weaker networks available

I can think of one advantage when always using the current best network: The White player can learn more efficiently if the Black player is learning too. However the Black player will not learn anything if it is an old network. Only after more network promotions, the Black player will slowly pick up some knowledge that has been learned from these games.

**Comment by Friday9i** [comment]
Nice initiative!
I think a different (and older) network would be more realistic for handi, but understand it's harder to program and harder to maintain...
Why would it be more realistic?
Because in handi games, the weaker player should not think about some good moves, because he just don't know them! That would happen using a weaker network but not with the current net and low playouts (1 playout "intuition" of the current network for black is already more than enough to beat high visits of the white player with 4 or more handi stones)... And low playouts can lead to ladder mistakes, which is bad. 
So unfortunately, I fear that asymmetric playouts would help for low handi (but LZ plays already low handi quite well) while not helping much for high handicap ; -(.

**Comment by fishcu** [comment]
Would it be possible to weaken one side by just adding more noise instead of decreasing the playouts so much?

For handicap training, komi should really be at 0.5! Even if it requires re-training the network.

**Comment by tapsika** [comment]
> For handicap training, komi should really be at 0.5! Even if it requires re-training the network.

I think the opposite: it is a good thing most bots play handicap games with white passes, and I hope this gets some popularity in human games as well. Even humans are better off if they only need to "train" for a single way of viewing or feeling the board (mentally adding W komi). I preferred this even before the computer go era.

On the practical side I think LZ should simply start playing a few % of selfplay with one initial white pass. This could even benefit non-handicap strength (more diversity).

**Comment by jkiliani** [comment]
The initial white pass(es) would have to be coupled with forbidding black to pass before the first white stone is played. By Tromp-Taylor rules, black could otherwise win those games without white making any move, which would not be the point of the handicap training.

**Comment by tapsika** [comment]
Sure, but that is easy to avoid. More important is the potential bias: if the net sees, say, 8% games where W makes a handi pass, it can start to think W's general chances are lower than in reality (since W will lose the majority of those games, thus lowering general W winrate). So maybe better to have it both direction: a few % where B passes first move, and another few % where W does the same. Also using ELF in those games may be doubtful.

**Comment by betterworld** [comment]
@tapsika 

> So maybe better to have it both directions: a few % where B passes first move, and another few % where W does the same.

So essentially I can think of two kind of biases that can be a problem in handicap training:

1. The Black player is probably weaker
2. In the opening the position is even if one player has passed

Your suggestion to alternate Black and White handicap would fix the first one. To fix the second one, I think it would suffice to make sure that there are more games where the game is lost by the one who passed.

**Comment by tapsika** [comment]
Some asymmetry remains even if both colors do some passes. B passing is not completely the same (W moves first and receives komi) as W passing (B moves twice and gives komi). I doubt this would cause problems but it also seems possible to eliminate.

**Comment by OmnipotentEntity** [comment]
Best candidate for handi 2 black fixed at 3200 visits so far is 30,000 visits for white, games are quite slow though:

```
players = {
    'leelaz-3200' : Player("./leelaz-next -m 30 -n -d -q -t 1 -r 3 --gpu 1 --noponder -v 3200 -w 10bc1042 --gtp", startup_gtp_commands=['time_settings 0 1 0']),
    'leelaz-30000' : Player("./leelaz-next -m 30 -n -d -q -t 1 -r 3 --gpu 0 --noponder -v 30000 -w 10bc1042 --gtp", startup_gtp_commands=['time_settings 0 1 0']),
}
```

Time settings are obligatory because otherwise the engine will stop searching too early to conform to a default 1 hour time block.

```
leelaz-3200 v leelaz-30000 (11/400 games)
board size: 19   handicap: 2 (fixed)   komi: 7.5
               wins                  avg cpu
leelaz-3200       5 45.45%   (black)  2970.27
leelaz-30000      6 54.55%   (white) 14741.36
```
